### PR TITLE
NF: Introduce List assertion

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
This way, it uses same tools as array equality and will show descriptive message about the difference

### CODE REVIEW

**Review by david-allison-1 — APPROVED** [review]
I'm fairly ambivalent about this, a Hamcrest matcher (or documentation on using an existing matcher) would likely be better from a readability standpoint. 

The syntax here seems verbose and copying the arrays is a negligible perf hit.

But, it's not incorrect, tests still pass.

**Comment by Arthur-Milchior** [comment]
> I'm fairly ambivalent about this, a Hamcrest matcher (or documentation on using an existing matcher) would likely be better from a readability standpoint.

Readability of what ?

I find that current error messages are hard to read, because it shows the whole list and not where the equality break. I changed this so that the error message we get shows either the change in length or the first difference in the list. So debugging is easier.

What do you think in terme of documentation ? I don't understand what could be done. However, I'd be fine creating a better test rule as long as it gives detailed explanation of the difference

> The syntax here seems verbose and copying the arrays is a negligible perf hit.

`ListUtil.assertListEquals` can be replaced by `assertListEquals` with a static import. Done in new version

**Comment by david-allison-1** [comment]
> Readability of what ?

of the code: `assertThat(expectedList, is(actualList), "message");` strikes me as much more readable, but this is personal preference.

> I find that current error messages are hard to read, because it shows the whole list and not where the equality break. I changed this so that the error message we get shows either the change in length or the first difference in the list. So debugging is easier.

Agreed

> What do you think in terme of documentation ? I don't understand what could be done. However, I'd be fine creating a better test rule as long as it gives detailed explanation of the difference

In terms of better explaining the testing mechanisms that we have: If we have Hamcrest's `is` or `contains` and they do the job, then we'd want to document that, as it's a very common operation, and the Hamcrest documentation is a little too general for my taste.

> `ListUtil.assertListEquals` can be replaced by `assertListEquals` with a static import. Done in new version

Definitely a good change

**Comment by Arthur-Milchior** [comment]
I rewrote it slightly so that it's even less verbose. It takes a sequence of expected values instead of a list

File: AnkiDroid/src/test/java/com/ichi2/utils/ListUtil.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * Asserts that two object lists are equal (same size and components in same order). If they are not, an
+     * {@link AssertionError} is thrown with the given message. It states "array" instead of list If
+     * <code>expecteds</code> and <code>actuals</code> are <code>null</code>,
+     * they are considered equal.
+     *
+     * @param actuals List of objects with actual values.
+     * @param expecteds List of values that should be in the list as is
+     */
+    public static <T>  void assertListEquals(List<T> actuals,
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/ListUtil.java:44]: I find this to be less readable. We're also going against the convention of `expected, actual`

**Review by david-allison-1 — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/test/java/com/ichi2/utils/ListUtil.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/test/java/com/ichi2/utils/ListUtil.java:44]: Then we can remove the second commit and just keep the first

**Review by david-allison-1 — APPROVED** [review]
Comments still stand, no point in bikeshedding on this

**Comment by Arthur-Milchior** [comment]
i'm fine with adding comments in the wiki page related to testing.

File: AnkiDroid/src/test/java/com/ichi2/utils/ListUtil.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package com.ichi2.utils;
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/utils/ListUtil.java:1]: missing copyright header

File: AnkiDroid/src/test/java/com/ichi2/testutils/AnkiAssert.java [hunk 3]
```diff
@@ -4,12 +4,14 @@
 
 import com.ichi2.libanki.Card;
 import com.ichi2.libanki.Collection;
+import com.ichi2.utils.ListUtil;
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/testutils/AnkiAssert.java:7]: if you import static you don't need the general import I think

File: AnkiDroid/src/test/java/com/ichi2/libanki/ModelTest.java [hunk 4]
```diff
@@ -5,6 +5,7 @@
 import com.ichi2.anki.exception.ConfirmModSchemaException;
 import com.ichi2.utils.JSONArray;
 import com.ichi2.utils.JSONObject;
+import com.ichi2.utils.ListUtil;
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/libanki/ModelTest.java:8]: if you import static you don't need the general import I think

File: AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java [hunk 5]
```diff
@@ -54,6 +54,7 @@
 import org.junit.Assume;
 import org.junit.AssumptionViolatedException;
 import org.junit.Before;
+import org.junit.Test;
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java:57]: I think all 3 imports added here are unneded?

**Review by mikehardy — CHANGES_REQUESTED** [review]
I have found array error messages difficult in the past as well
This solution seems fine - I think the actual readability of the in-test code is fine in this version and if it produces better error messages, okay by me

I left nitpick comments though :)

**Comment by Arthur-Milchior** [comment]
Done

File: AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java (hunk 5 above)
Comment (mikehardy) [reply AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java:57]: @Arthur-Milchior I think this one is still unused, but other than that it's all patched up

**Review by mikehardy — CHANGES_REQUESTED** [review]

**Comment by Arthur-Milchior** [comment]
done

**Review by mikehardy — APPROVED** [review]
good to go!

**Pull request merged**